LFS-680 + LFS-681: When clicking (+) on a subject child chart to add a form, do not offer irrelevant subjects and questionnaires

OrdiNeu opened the pull request:
### [LFS-680](https://phenotips.atlassian.net/browse/LFS-680): When clicking (+) on a subject chart to add a form, provide list of forms that take this type of subject and its children only

This work also includes the ProgenyProcessor, which allows for a new .json selector (.progeny) which recursively finds the children of a given node (via recursively looking for nodes whose parent is the given node).

To test: When viewing the patient chart for a Tumour, trying to create a new form from that page will no longer show Questionnaires like Demographics, which cannot be used for a Tumour.

### [LFS-681](https://phenotips.atlassian.net/browse/LFS-681): When clicking (+) on a subject child chart to add a form, do not show parent subject in the subject selection dialog

**To test**: When creating a new form **from a patient chart**, the following behaviours should occur:
- If you try to create a form whose subject can be the patient you're looking at, it automatically chooses the patient as the subject for your new form. E.g., create a new patient, go to its chart, and then try creating a Demographics form. It should automatically choose the patient you're looking at as the subject for that form.
- If you try to create a form for a new subject, and the parent of that subject can be the patient you're looking at, the patient is automatically chosen as its parent. For example, create a Patient, then go to its chart and create a Tumour form. It will prompt you to create a new Tumour, which should then **automatically* choose the patient you're looking at as its parent.

sashaandjic commented:
I do not expect to see other subject's children of the same type. This perhaps will be addressed in [871](https://phenotips.atlassian.net/browse/LFS-871)
![image](https://user-images.githubusercontent.com/23243244/105789546-a5495100-5f50-11eb-832c-313be20dfac3.png)

OrdiNeu commented:
> We should still be showing the eligible children.
> ...
> It works well when I open tumor T's chart and create a Radio form. The Radio form gets created for the tumor T.

This has been fixed. in `069f398` :heavy_check_mark:

sashaandjic commented:
The parent is still shown in the **Select a subject** dialog
![image](https://user-images.githubusercontent.com/23243244/106200552-28eb8380-6185-11eb-83a7-380dbd84d096.png)

Steps to reproduce:
- create patient and tumor
- go to tumor chart
- click on + and choose Radio form
- **Select a subject** dialog opens up offering patient and tumor
- ideally the Radio form should just be created for the tumor shown in the tumor chart, rather than offering the same tumor as a singe option in the dialog
![image](https://user-images.githubusercontent.com/23243244/106200469-078a9780-6185-11eb-94f5-7eaeba27ddc0.png)

OrdiNeu committed:
When clicking (+) on a subject child chart to add a form, do not show
parent subject in the subject selection dialog

OrdiNeu committed:
LFS-681: When clicking (+) on a subject child chart to add a form, do
not show parent subject in the subject selection dialog

Bugfix: Remove empty ancestry/progeny properties when they no longer exist.

OrdiNeu committed:
[Misc] Fix broken references in the Variants Questionnaire

OrdiNeu committed:
LFS-681: When clicking (+) on a subject child chart to add a form, do
not show parent subject in the subject selection dialog

Fix the questionnaire editor

OrdiNeu committed:
LFS-681: When clicking (+) on a subject child chart to add a form, do
not show parent subject in the subject selection dialog

Switch code away from using Progeny
Use a recursive function to find progeny subject types, instead of the
default.

OrdiNeu committed:
LFS-680: When clicking (+) on a subject chart to add a form, provide
list of forms that take this type of subject and its children only

sashaandjic commented:
Found only one inconsistency. Steps to reproduce:
- create patient and tumor
- go to the tumor chart
- click on (+)
- select Radiotherapy form
- click on the NEW button
- expected outcome: only subject tumor type should be shown (which upon creation should be assigned to the current patient)
- actual outcome: both patient and tumor types are offered in the dialog 

![image](https://user-images.githubusercontent.com/23243244/111390920-ef65cd80-8689-11eb-80cd-3ea6f91b9b1d.png)

marta- commented:
> Found only one inconsistency. Steps to reproduce:
> 
> * create patient and tumor
> * go to the tumor chart
> * click on (+)
> * select Radiotherapy form
> * click on the NEW button
> * expected outcome: only subject tumor type should be shown (which upon creation should be assigned to the current patient)
> * actual outcome: both patient and tumor types are offered in the dialog

That's a bit out of scope, and should be covered by https://phenotips.atlassian.net/browse/LFS-877 .

sashaandjic reviewed (approved):
tested

acrowthe reviewed (approved):
Looks good
